tests/test_enum.rs

```diff
@@ -0,0 +1,50 @@
+use pyo3::prelude::*;
+use pyo3::{py_run, wrap_pyfunction};
+
+mod common;
+
+#[pyenum]
+#[derive(Debug, PartialEq, Clone)]
+pub enum MyEnum {
```

Comment:
Crazy idea, does this need to be a separate `#[pyenum]` macro?

This still creates a Python class, so maybe this can just be a new feature for the `#[pyclass]` macro?

Reply:
I think that's a great idea. I'm not sure how do that interact with [https://pyo3.rs/v0.15.0/class.html#customizing-the-class] though. Maybe we just disable `extends`, `subclass` and `dict`? I'll implement this PR as a `#[pyclass]` if you are okay with it.

Reply:
Implementing as `#[pyclass]` sounds great to me!

> Maybe we just disable extends, subclass and dict?

I think `extends` makes sense to disable, however `subclass` and `dict` could reasonably be things that users want to add to customize their enum? I don't think they'll add much complexity to this PR, we can see as we go and decide that later.

Reply:
I just want to be future-proof. In the future when we add ADT support, each variant will be represented by a subclass (I explained why in [another comment](https://github.com/PyO3/pyo3/issues/417#issuecomment-972623184) of #417), and we want them to be the only subclasses of this enum, otherwise pattern matching will not work. (Python interpreter cannot recognize those subclasses as a variant of the enum.)

I am now okay with enabling `dict`, as I cannot think of a potential reason why we would wan't to disable `dict` in the future.

Reply:
> and we want them to be the only subclasses of this enum

Ahh good point, ok yes let's disable `subclass` too.

Reply:
Then we just need to figure out how to prevent subclassing from the Python side 😕.

Reply:
Worth writing a test but I'm fairly sure that neither Python nor Rust can subclass without `subclass`.

Reply:
Well, I just want to stop user from writing `#[pyclass(subclass)]`. We still need to secretly enable that in order to write our variant subclasses.

Reply:
I see. I'm not sure it can work like that; as far as Python is concerned either it's legal to subclass this or it isn't.

Might be able to control this a bit via a metaclass? Could be overkill.

Reply:
Python's `Enum` also makes use of metaclass to make sure its enums are immutable. I'll try to implement that after I finish `__eq__`.

Reply:
I'm sorry, for some reasons I thought `PyClass::BaseNativeType` must be a class created by the metaclass ... I found I was wrong after reading `create_type_object` and a bunch of CPython code... 😅😅
But I still have no idea whether it's possible create metaclass in PyO3, and if so what are the limitations. Did any one else try this before? Maybe I can do some experiments...

Reply:
Indeed there isn't yet a way to set metaclasses - #906.

Agreed, let's complete this PR without solving the metaclass issue and we can always explore that as a follow-up.